File: python/백준/부르트 포스/17626 Four Squares.py
import math

n = int(input())
d = [0] * 50001

def solution(n):
    if int(math.sqrt(n)) ** 2 == n:
        print(1)
        exit(0)
        
    for i in range(1, int(math.sqrt(n)) + 1):
        if int(math.sqrt(n - i * i)) ** 2 == (n - i * i):
            print(2)
            exit(0)
    
    for i in range(1, int(math.sqrt(n)) + 1):
        for j in range(1, int(math.sqrt(n - i * i)) + 1):
            if int(math.sqrt(n - i * i - j * j)) ** 2 == (n - i * i - j * j):
                print(3)
                exit(0)
    
    print(4)
    
solution(n)
    

File: cpp/백준/부르트 포스/17626 Four Squares.cpp
#include <iostream>
#include <cmath>
using namespace std;

int n;

void solution(int n) {
    int r = (int)sqrt((double)n);
    if (r * r == n) {
        cout << 1 << "\n";
        exit(0);
    }

    for (int i = 1; i <= (int)sqrt((double)n); ++i) {
        int rem = n - i * i;
        int s = (int)sqrt((double)rem);
        if (s * s == rem) {
            cout << 2 << "\n";
            exit(0);
        }
    }

    for (int i = 1; i <= (int)sqrt((double)n); ++i) {
        for (int j = 1; j <= (int)sqrt((double)(n - i * i)); ++j) {
            int rem = n - i * i - j * j;
            int s = (int)sqrt((double)rem);
            if (s * s == rem) {
                cout << 3 << "\n";
                exit(0);
            }
        }
    }

    cout << 4 << "\n";
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    cin >> n;
    solution(n);
    return 0;
}
